### PR TITLE
Feat/focus trap specified

### LINKED ISSUE #2190: Have focusTrap action focused on a specific element using tabIndex
### Describe the feature in detail (code, mocks, or screenshots encouraged)

After discussing this on discord with @endigo9740 :
Let users implement their own tabindex attributes, so they retain full control and encourage best practices
By default, focusTrap should default to the first focusable element in the array
If specified, focusTrap should target the tabindex supplied by the user (I'll discuss how we pass this below)
Update modals/dialogs to accept an optional tabindex, which is passed to the embedded version of focusTrap. Which triggers the above result.
Per the interface for providing the tabindex to the focusTrap, we're unfortunately a bit limited due to the current implementation accepting only a single boolean value:
`<form use:focusTrap={someBoolValue}>`

First instinct is to do something like this - however this would be a breaking change and is not possible until the next major release:
`<form use:focusTrap={{state: someBoolValue, tabindex: 2}}>`

But what we'll likely need to do is use a data attribute as a temporary solution. This is optional and non-breaking:
`<form use:focusTrap={someBoolValue} data-tabindex={2}>`

I'm demonstrating tabindex as a number here, but string might be more appropriate.

We should likely keep the existing logic around a default selection (first focusable). This keeps parity with the existing functionality (prevents a breaking change again), as well as provides a nice fallback if users don't specify tabindex values in their content as expected.

### What type of pull request would this be?

Enhancement

### Provide relevant links or additional information.

_No response_

### PR DESCRIPTION
**Pull request by doic**
## Linked Issue

Closes #2190

## Description

The goal is to allow focusTrap to focus on a specified element. If no element is specified, focusTrap will focus on the first focusable element, taking into consideration the tabindex attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

Focus specified elements are targeted thanks to their data-tabindex attribute, which don't conflict with the tabindex attribute.

## Changsets

Instructions: Changesets automate our changelog. If you modify files in `/packages/skeleton`, run `pnpm changeset` in the root of the monorepo, follow the prompts, then commit the markdown file. Changes that add features should be `minor` while chores and bugfixes should be `patch`. Please prefix the changeset message with `feat:`, `bugfix:` or `chore:`.

## Checklist

Please read and apply all [contribution requirements](https://www.skeleton.dev/docs/contributing).

- [X] This PR targets the `dev` branch (NEVER `master`)
- [X] Documentation reflects all relevant changes
- [X] Branch is prefixed with: `docs/`, `feat/`, `chore/`, `bugfix/`
- [X] Ensure Svelte and Typescript linting is current - run `pnpm check`
- [X] Ensure Prettier linting is current - run `pnpm format`
- [X] All test cases are passing - run `pnpm test`
- [X] Includes a changeset (if relevant; see above)

### CODE REVIEW

**Comment by endigo9740** [comment]
Thanks @doic everyone looks good with a quick scan. I tend to do proper PR reviews on Fridays, so expect at least at least some preliminary feedback tomorrow. When you're ready for a proper review, tap the "ready for review" option here. Thanks!

**Comment by doic** [comment]
Done. I had run pnpm check before messing with the Modal, sorry :-(

**Comment by endigo9740** [comment]
@doic so a few issues here -

1. `svelte-check` appears to still be in a failing state. Check the UI just below this comment for details.
2. Supplying a `tabIndex` in `drawerSettings` or `modalSettings` seems to work, but only if it's a value greater than `0`. Remember Javascript treats `0` as a falsey state, so check your conditional logic.
3. Supplying `tabIndex` as a prop does not appear to be working at all for the either the Drawer or Modal components. I'd recommend checking the flow the data to locate where this is failing
4. I have some suggestions per the documentation (both page and JSDoc comments), but we can return to these after the functional issues are sorted.

Per issues 2 and 3 above, this can be tested by inserted focusable elements with `data-tabindexes` that are presented out of numeric order:

```
<input type="text" placeholder="test-1" data-tabindex={1} />
<input type="text" placeholder="test-2" data-tabindex={0} />
<input type="text" placeholder="test-3" data-tabindex={2} />
```

- When you set `data-tabIndex={0}`, I expect it to select the second item, but instead it selects the first
- When you set `data-tabIndex={2}`, it will correctly select the third item
- When you set `tabIndex: 0`, I expect it to select the second item, but instead it defaults to the first
- When you set `tabIndex: 2`, I'd expect it to select the third item, but instead it defaults to the first

One other quick suggestion - making the value of `data-tabindex` a string might be more familiar to folks, to better match the native iteration, ex:`tabindex="3"`. Not totally required, but it may also fix the issue with `0` being treated as a falsey statement.

**Comment by endigo9740** [comment]
One more quick ask - can you put a space between `feat:` and the description in the changelog? This is the line item text that will be inserted into our auto-generated changelog. So it'll be provided verbatim:

Screenshot 2023-10-27 at 12 24 17 PM

**Comment by doic** [comment]
Weird. Check command worked on my machine. I'll check everything tomorrow morning. Thanks for the feedback 👍

**Comment by doic** [comment]
OK - I had run pnpm check in packages dir, but not in svelte.dev dir. Makes sense now 😅.
I did choose to work with numbers because it felt more practical for some logic, like:
```typescript
someBool ? tabIndex : tabIndex + 1
```
Obviously, this makes tabIndex = 0 falsey, which is the perfect default value.
But it makes impossible to target a data-tabindex="0" 😑
If you think working with string is the best, let's do this.
I didn't take into consideration passing tabIndex to Modal and Drawer components as a prop, I will work on it !

**Comment by doic** [comment]
By the way, the more I see `tabindex` and `data-tabindex` HTML attributes side by side, the more I think using `tabIndex` as a parameter for this feature is wrong. It can be very confusing.
It's not an index, and it has nothing to do with TAB...
Can we find a better name ? I think "focus + target" sounds good 😁

File: packages/skeleton/src/lib/actions/FocusTrap/focusTrap.ts [hunk 1]
```diff
@@ -21,16 +24,30 @@ export function focusTrap(node: HTMLElement, enabled: boolean) {
 		}
 	}
 
+	// Sort focusable elements by tabindex, positive first, then 0
+	const sortByTabIndex = (focusableElems: HTMLElement[]): HTMLElement[] => {
```
[the remark this answers is not in the document]
Comment (doic) [reply packages/skeleton/src/lib/actions/FocusTrap/focusTrap.ts:28]: It's different. We need 0 indexes AFTER positive indexes, and we need to filter out negative indexes. That's how tabindex attributes are used, and that's exactly what the function does :-)

[the remark this answers is not in the document]
Comment (doic) [reply packages/skeleton/src/lib/actions/FocusTrap/focusTrap.ts:28]: To be clear, that's not really part of the "focus target" idea; it was the first idea I had, and the reason why this feature was named after "tabindex" 😅.
At the beginning, I was just looking at a simple way to influence the target of the focus by using the tabindex attribute.
So, just by adding `tabindex="1"` to my `<input />`  was enough to focus on it.
Then the idea evolved to something more specific, and unrelated to the `tabindex` attribute, but still, I think we should take the tabindex information into consideration when falling back to the default behavior (focusing on the smallest positive tabindex value > 0)

[the remark this answers is not in the document]
Comment (Mahmoud-zino) [reply packages/skeleton/src/lib/actions/FocusTrap/focusTrap.ts:28]: @doic 
My bad I miss-interpreted the code, but this is an indication that the code in this function is not clean.
my suggestion would be something like this:
```js
return focusableElems.filter(elem => elem.tabIndex >= 0).sort((a, b) {
    if (a.tabIndex === 0 && b.tabIndex > 0) return 1; // Move 0 to end of array
    else if (a.tabIndex > 0 && b.tabIndex === 0) return -1; // Move 0 to end of array
    else return a.tabIndex - b.tabIndex; // Sort non-zero values in ascending order
}));
```
this is still better than spreading and copying the array after filtering and should achieve the same results.

Please note I didn't had the chance to test the code provided here so you may have to change some stuff to make it work as expected.

Also please consider adding comments whenever you feel something is not 100% clear, other devs will have to read and possibly change/extend your code.

[the remark this answers is not in the document]
Comment (doic) [reply packages/skeleton/src/lib/actions/FocusTrap/focusTrap.ts:28]: You're filtering on tabIndex >= 0 then checking if tabIndex === 0
I understand you didn't check your code, nor tested it, no worry 😊
I also want to produce the best code, the most performant one.
I'm going to compare some scenari and see if one is better.
I'll add more comments; you're right they're needed here 👍

[the remark this answers is not in the document]
Comment (doic) [reply packages/skeleton/src/lib/actions/FocusTrap/focusTrap.ts:28]: I'll see how to improve this, thanks for the feedback 🙏

**Comment by endigo9740** [comment]
Hey @doic I handle PR reviews on Sundays now. Just gave this a quick scan. I see you received some feedback from Mahmoud, but I don't see any commits beyond that. I'll try to squeeze in another review on Tuesday (our next release) otherwise expect a follow up later in the week.

**Comment by doic** [comment]
Yes sorry, I've been quite busy these days. I'll try to get back to it during the week.

**Comment by endigo9740** [comment]
Hey @doic - it's my next round of reviews so just checking in. I think this is a positive change and would like to see it added, but we may have to put a deadline on this PR if we ever wish to see it added. If this is still pending for my next review (in 2 weeks) I'll likely go ahead and close this out. But will welcome you to resubmit at any time.

Not trying to rush you, I know how busy life can be, but don't want our pending PRs to languish forever ;)

**Comment by doic** [comment]
Hi @endigo9740 !
Yes I want to do it soon. Life and stuff.
I'm still new at PR - I'll need to find out the correct way to sync the current branch to mine - I guess there were a few changes since my fork.
I'll be busy on Monday but will try my best to get back at it on Tuesday 🙏

**Comment by doic** [comment]
> By the way, the more I see `tabindex` and `data-tabindex` HTML attributes side by side, the more I think using `tabIndex` as a parameter for this feature is wrong. It can be very confusing. It's not an index, and it has nothing to do with TAB... Can we find a better name ? I think "focus + target" sounds good 😁

I'm back 😁
Nobody answered this comment - can someone please give some feedback ?
After reading a bit more about other ways to handle this, I found this `createOptions.initialFocus` idea:
https://www.npmjs.com/package/focus-trap
I'm not talking about the whole implementation, only the name of it.
I find it easier to understand - `initialFocus` / `data-initialfocus` instead of `tabindex` / `data-tabindex`.

**Comment by doic** [comment]
The easiest way - imo - to do this would be with css selectors (allowing to use id attributes if we prefer):
```svelte
<form use:focusTrap={{ enabled: true, initialFocus: 'input[type=email]' }}>
	<input type="text" placeholder="Name" />
	<input type="email" placeholder="Email" />
	<button class="btn variant-filled-primary">Submit</button>
</form>
```

**Comment by endigo9740** [comment]
@doic I'll plan to do PR reviews tomorrow. I'll dive into this then.

**Comment by endigo9740** [comment]
@doic a lot of the original ideas were trying to replicate a proper tabindex-like system, something that would handle fallbacks and conditional statements where the focusable element might change as state updates.

Imagine you had a conditional statement inside a Drawer or Modal content like this:

```html
<!-- drawer contents -->
<div>
    {#if someBool}
        <button data-tabindex="0">Button 1</button >
    {#else}
        <button data-tabindex="1">Button 2</button >
    {/if}
    <button data-tabindex="2">Button 3</button >
</div>
```

Based on the state of `someBool` either Button1 or 2 will be the top-most focusable element, but that will differ based on state that may or may not change before this UI is rendered to the screen. With the `data-tabindex` system we can handle this easily. Including any scope of complexity (multiple conditionals, etc).

However, I do agree the naming is not quite semantic. As we're appending attributes purely meant to be used by the focus trap. As such, I'd recommend we do keep the same goal with functionality, but provide more semantic naming.

I tend to favor shorter, single word values whenever possible. Then let JSDocs provide more context via Intellesense. So I'd recommend something like:

* `initialFocus` -> `focus`
* `data-initialfocus` -> `data-focus` (or `data-focusindex`)

Let me know if that makes sense.

**Comment by doic** [comment]
@endigo9740, in your example, we'll have either 0 or 1 as an available `data-focusindex`.
So what focus value should we pass to the `focusTrap` action ?

We currently have something that behaves like this:
image

But I sense (in your example) that the `focus`/`tabIndex` value passed to the `focusTrap` action is not needed anymore, because the action has to determine what is the smallest `data-focusindex` value within the element.

If I'm correct, then this feels very tabindex-like and easier to use. We just need to use `tabindex` and/or `data-focusindex` and let `focusTrap` decide which element to focus on.

It will be more consistent with the previous `focusTrap` signature - no need to pass a `boolean | object` parameter as I was doing.

**Comment by endigo9740** [comment]
> But I sense (in your example) that the focus/tabIndex value passed to the focusTrap action is not needed anymore, because the action has to determine what is the smallest data-focusindex value within the element.

Yeah sorry, two different thoughts there that bled together. Yeah if we have the index model we go for the smallest index and a specified value is not required.

> If I'm correct, then this feels very tabindex-like and easier to use. We just need to use tabindex and/or data-focusindex and let focusTrap decide which element to focus on.

Yep, exactly this.

**Comment by Mahmoud-zino** [comment]
@doic the code looks good to me, I would just suggest changing the description in the docs to something like this:

----

## Focus order

Specify the order of focusable elements by assigning a value to the `data-focusindex` attribute. The element with the lowest value will be focused first.

**Review by Mahmoud-zino — APPROVED** [review]

**Comment by doic** [comment]
@Mahmoud-zino done 🙏

**Comment by endigo9740** [comment]
@doic I know this issue is waiting on me, but my schedule has blown up lately. I have some free time coming up in a couple weeks, so worst case I'll jump on this again then. Thanks so much for your patience, I'm very excited to get this implemented when possible!

**Comment by doic** [comment]
@endigo9740 sure, no problem :-) I hope this will make it to one of the next releases !

**Comment by endigo9740** [comment]
@doic I'm actually doing PR reviews as we speak today! So far we're down from 14 - 5. So I'm slowly inching my way in this direction :)

**Comment by endigo9740** [comment]
@doic with the smaller PRs sorted today I'm going to aim to jump on yours and the few other remaining tomorrow post-release. Your PR is a bit more broad in scope so I want to make sure I devote enough time to testing, rather than to rush it in last minute.

Expect some updates on this tomorrow. If things are close enough I may go ahead and make a few modifications myself direction to help wrap this up. But keep an eye out for my updates.

**Comment by endigo9740** [comment]
@doic @Mahmoud-zino I've pushed some very minor changes to the documentation info and examples. If you two sign off on this I'm happy to merge. The feature is working exactly as expected. Great job!

File: sites/skeleton.dev/src/routes/(inner)/actions/focus-trap/+page.svelte
Comment (doic) [inline sites/skeleton.dev/src/routes/(inner)/actions/focus-trap/+page.svelte:1]: The description looks better, indeed, but I don't understand why hardcoding `use:focusTrap={true}` line 104 ?
I basically copied the example above - where `use:focusTrap={isFocused}` (line 54)

**Review by endigo9740 — APPROVED** [review]

File: sites/skeleton.dev/src/routes/(inner)/actions/focus-trap/+page.svelte
Comment (endigo9740) [reply sites/skeleton.dev/src/routes/(inner)/actions/focus-trap/+page.svelte:1]: In the first example we're teaching folks how to toggle the focus trap on and off. No need to repeat that in the second. The area of focus is `focusindex` values.

**Review by Mahmoud-zino — APPROVED** [review]

**Pull request merged**